src/eval/merge.rs

```diff
@@ -426,6 +427,24 @@ pub fn merge(
                 env,
             })
         }
+        // We want to merge a non-record term with a record contract
+        (t1_, t2_ @ Term::Record(..)) => {
+            if let MergeMode::Contract(label) = mode {
+                Err(EvalError::BlameError(label, call_stack.clone()))
+            } else {
+                Err(EvalError::MergeIncompatibleArgs(
+                    RichTerm {
+                        term: SharedTerm::new(t1_),
+                        pos: pos1,
+                    },
+                    RichTerm {
+                        term: SharedTerm::new(t2_),
+                        pos: pos2,
+                    },
+                    pos_op,
+                ))
+            }
+        }
```

Comment:
I think you can reuse the case below with an `if` guard, instead of duplicating the code creating the `MergeIncompatibleArgs`. There are several solutions to nest the matching, but here is an example:

```suggestion
        (t1_, t2_ @ Term::Record(..)) if matches!(mode, MergeMode::Contract(..)) => {
            if let MergeMode::Contract(label) = mode {
                Err(EvalError::BlameError(label, call_stack.clone()))
            } else {
                unreachable!()
            }
        }
```

Another solution is to keep the generic match `(t1_, t2_)` and then match again, which saves the redundant match on `mode` and the `unreachable!()` at the price of a nested pattern matching:

```rust
(t1_, t2_) => {
    match ((mode, t2_)) {
        (MergeMode::Contract(label), Term::Record(..)) => Err(EvalError::BlameError(label, call_stack.clone())),
        _ => {
            Err(EvalError::MergeIncompatibleArgs(
                RichTerm {
                    term: SharedTerm::new(t1_),
                    pos: pos1,
                },
                RichTerm {
                    term: SharedTerm::new(t2_),
                    pos: pos2,
                },
                pos_op,
            ))
        }
    }
}
```

Yet another solution is to make a local function `mk_non_mergeable_error` or the like.

Reply:
I don't like the `unreachable!()`, I'll go with the later.